desc/experimental/_terpsichore.py

```diff
@@ -0,0 +1,872 @@
+            surf_idx = int(row[0])
+            line_num = line_num + 1
+            if surf_idx >= surfs - 1:
+                break  # NOTE: not including vacuum surfaces
+    file.close()
+
+    errorif(len(growth_rate) == 0, RuntimeError, "Growth rate not found!")
+    errorif(
+        len(growth_rate) > 1,
+        NotImplementedError,
+        "Not capable of handling multiple growth rates.",
+    )
+    errorif(len(delta_W) != surfs - 1, RuntimeError, "Incorrect number of dW values!")
+
+    return growth_rate[0] if scalar else delta_W
+
+
+class TERPSICHORE(ExternalObjective):
+    """Computes ideal MHD linear stability from calls to the code TERPSICHORE.
+
+    Returns the linear growth rate of the fastest growing instability, or the ΔW values
+    at each flux surface. A positive growth rate or negative ΔW denotes instability.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    abs_step : float, optional
+        Absolute finite difference step size. Default = 1e-4.
+        Total step size is ``abs_step + rel_step * mean(abs(x))``.
+    rel_step : float, optional
+        Relative finite difference step size. Default = 0.
+        Total step size is ``abs_step + rel_step * mean(abs(x))``.
+    processes : int, optional
+        Maximum number of CPU threads to use for multiprocessing. Default = 1.
+    path : str, optional
+        Path to the directory where temporary files will be stored.
+    exec : str, optional
+        File name of the TERPSICHORE executable. Must be located in the directory
+        specified by ``path``.
+    scalar : bool, optional
+        If Ture, returns the growth rate of the fastest growing instability (default).
+        If False, returns the ΔW values at each flux surface.
+    mode_family : int, optional
+        The mode family of the instabilities to consider. Only implemented for NFP =
+        2 or 3, which only have two mode families: N=0 and N=1. If ``mode_family < 0``,
+        both mode families are considered. Default = -1.
+    surfs : int, optional
+        Number of surfaces to include in the equilibrium input. More surfaces provides
+        more accuracy at the cost of longer compute times. Default = 101.
+    lssl : int, optional
+        Minimum number of possible permutations of Boozer mode combinations
+        (determined by ``M_booz_max`` and ``N_booz_max``). If TERPSICHORE fails to run,
+        try increasing this parameter. Default = 1000.
+    lssd : int, optional
+        Minimum number of possible permutations of stability mode combinations
+        (determined by ``M_max`` and ``N_max``). If TERPSICHORE fails to run,
+        try increasing this parameter. Default = 1000.
+    M_max : int, optional
+        Maximum poloidal mode number of stability modes to consider. Will include modes
+        with m ∈ [0,M_max] (if ``mode_family < 0``). Default = 8.
+    N_min : int, optional
```

Comment:
is this mode number the mode structure of the instability within one field period?